mono/utils/mono-counters.c

```diff
@@ -518,6 +518,14 @@ mono_counters_sample (MonoCounter *counter, void *buffer, int buffer_size)
 	return sample_internal (counter, buffer, buffer_size);
 }
 
+// We want to default to g_print if outfile is not specified. This matters because on some platforms (e.g. Android),
+// printing to stdout is not the actual way to see output, and we've wrapped that logic in `g_print`.
+#define FPRINTF_OR_G_PRINT(outfile, ...) if (outfile) { \
```

Comment:
Please make this a function instead of a macro.
There is no reason for it to be a macro.
You want the code to be big and slow or small and slow?

Reply:
Also try to think of another name.
Maybe mono_fprintf, with the understanding somehow implied, that outfile==null has special handling??

Reply:
Btw, I was going to say use do/while around the macro, but it just does not need to be a macro.
You should always question why something is a macro and favor functions.

Reply:
Maybe mono_printerrf? Are these always for errors?

Reply:
I really don't think this should be used more broadly in the codebase, which is why I'm disinclined to prefix it with `mono_`. The macro is pretty ugly, agreed there - I think my thought process was that I could easily undef it to ensure it wasn't used elsewhere. Probably better to just use a static function.

The decision to keep it local may have been short-sighted if there are other cases where we have functions taking in and using streams, but I'm not sure if we do. I looked through the code a bit last night and the situation is a bit of a mess otherwise, see #18580.

Reply:
I think everything might as well start `mono_`.
Then you can add/remove `static` as needed..
If a good name can be had, then no need for undef, even if not broadly used.

Lambdas are good for local functions too..

I admit I wrote something similarly bad today:

```
#define PRINT(x) fputs((x), stdout); OutputDebugStringA(x)
        PRINT("In debugger, gn, not g or gh, for this to run correctly.\n");
```
used once in a small file that houses just a test,we'll see how it flies (my team is *adamantly* against macros...)

At the very least, in future, avoid macros. Favor functions.

Reply:
This counters stuff is going to go through an overhaul Real Soon - both Android and WebAssembly would actually prefer a callback API to collect the counter data so that you don't have to look in the adb log or in the browser console for this data.  (And on an even longer term this will get wired up to EventSource once we have that going).  just printing to a `FILE*` is the wrong interface here.